### PR TITLE
Callback curriculum

### PR DESCRIPTION
**Pull request by TheSpaceSheep**
The curriculum is now managed by the callback system of EGG.
The next task is not given by an accuracy threshold anymore but by detecting plateaus via a double running mean.
The optimizer resets at each curriculum step.

### CODE REVIEW

File: egg/core/util.py [hunk 1]
```diff
@@ -226,6 +226,13 @@ def build_optimizer(params: Iterable) -> torch.optim.Optimizer:
     return optimizer(params, lr=get_opts().lr)
 
 
+def reset_optimizer_state(module: torch.nn.Module, optimizer: torch.optim.Optimizer):
+    opt_state_dict = optimizer.state_dict()
+    for name, _ in module.named_parameters():
+        opt_state_dict["state"][name] = []
+    optimizer.load_state_dict(opt_state_dict)
```
Comment (pmichel31415) [inline egg/core/util.py]: Add a quick docstring

File: egg/zoo/compo_vs_generalization/callbacks.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import egg.core as core
+from egg.core.interaction import Interaction
+from egg.core.util import reset_optimizer_state
+from egg.zoo.compo_vs_generalization.curriculum_games import CurriculumGameWrapper
+
+import torch
+import numpy as np
+
+
+class CurriculumManager(core.Callback):
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/callbacks.py]: The name is a bit too generic for what you are doing (you could manage the curriculum in a different way).

How about making the name more explicit and adding a quick docstring to explain the strategy as well as the arguments?

File: egg/zoo/compo_vs_generalization/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, 
+            game: CurriculumGameWrapper, 
+            optimizer: torch.optim.Optimizer,
+            acc_threshold=None):
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/callbacks.py]: This can be annotated as `Optional[float]`

File: egg/zoo/compo_vs_generalization/callbacks.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+        self.game = game
+        self.optimizer = optimizer
+        self.acc_threshold = acc_threshold
+        self.acc_ors = []
+
+    def on_epoch_end(self, loss: float, logs: Interaction, epoch: int):
+        acc_or = logs.aux["acc_or"].mean().item()
+        update_curriculum = False
+        if self.acc_threshold is None:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/callbacks.py:40]: The fact that acc_threshold == None corresponds to a specific behaviour should be explained somewhere in a docstring

File: egg/zoo/compo_vs_generalization/callbacks.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            self.acc_ors.append(acc_or)
+            if len(self.acc_ors) > 20:
+                del self.acc_ors[0]
+
+                # compute running averages
+                prev_run_avg = np.array(self.acc_ors[:10]).mean()
+                next_run_avg = np.array(self.acc_ors[10:]).mean()
+                print(abs(prev_run_avg - next_run_avg)/prev_run_avg)
+
+                if abs(prev_run_avg - next_run_avg)/prev_run_avg < 1e-4:
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/callbacks.py]: Spacing between operators is inconsistent. Consider using autopep8 or other automatic formatting plugin

File: egg/zoo/compo_vs_generalization/callbacks.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        update_curriculum = False
+        if self.acc_threshold is None:
+            self.acc_ors.append(acc_or)
+            if len(self.acc_ors) > 20:
+                del self.acc_ors[0]
+
+                # compute running averages
+                prev_run_avg = np.array(self.acc_ors[:10]).mean()
+                next_run_avg = np.array(self.acc_ors[10:]).mean()
+                print(abs(prev_run_avg - next_run_avg)/prev_run_avg)
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/callbacks.py]: Is this for debugging?

File: egg/zoo/compo_vs_generalization/curriculum_games.py [hunk 7]
```diff
@@ -38,8 +38,6 @@ def __getattr__(self, attrname):
             return getattr(self.game, attrname)
             
 
-
-
```
Comment (pmichel31415) [inline egg/zoo/compo_vs_generalization/curriculum_games.py:42]: Some weird and inconsistent line spacing here too. Again, consider autoformatting to handle this (if you plan to run autoformat do it via a separate PR to reduce clutter)

**Review by pmichel31415 — CHANGES_REQUESTED** [review]
Looks good, although the curriculum callback could be streamlined a bit, see comments. There are also a number of formatting inconsistencies, consider using autoformatting tools in the future to harmonize things

**Review by pmichel31415 — APPROVED** [review]
LGTM!
